server/src/main/java/org/elasticsearch/cluster/routing/OperationRouting.java

```diff
@@ -125,16 +126,32 @@ public GroupShardsIterator<ShardIterator> searchShards(
                 nodeCounts
             );
             if (iterator != null) {
-                var shardsThatCanHandleSearches = iterator.getShardRoutings()
-                    .stream()
-                    .filter(shardRouting -> canSearchShard(shardRouting, clusterState))
-                    .toList();
+                final List<ShardRouting> shardsThatCanHandleSearches;
+                if (isStateless) {
+                    shardsThatCanHandleSearches = statelessShardsThatHandleSearches(clusterState, iterator);
+                } else {
+                    shardsThatCanHandleSearches = statefulShardsThatHandleSearches(iterator);
+                }
                 set.add(new PlainShardIterator(iterator.shardId(), shardsThatCanHandleSearches));
             }
         }
         return GroupShardsIterator.sortAndCreate(new ArrayList<>(set));
     }
 
+    private static List<ShardRouting> statefulShardsThatHandleSearches(ShardIterator iterator) {
+        final List<ShardRouting> shardsThatCanHandleSearches = new ArrayList<>(iterator.size());
+        for (ShardRouting shardRouting : iterator) {
+            if (shardRouting.isSearchable()) {
```

Comment:
This check seems like it shouldn't even be needed in stateful but we have a hacky test that depends on it so I left it, doesn't matter much anyway since the setting check was the bulk of the runtime cost here.